---
title: "gvis_atac"
output: html_notebook
---

# Introduction


# install packages

Here we install Gviz and their dependencies to allow for genome visualization. Additional packages such as `BSgenome` and `TxDb` contain sequence information and genomic coordinate info respectively. Specifically, we install the hg19 version of these packages, since we know we will be using hg19 aligned ATAC-seq track and signal files. Typically hg38 is preferred for newer data analysis, but it is important to note the genomic coordinates will differ between hg19 and hg38.

```{r installation, eval = FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("Gviz")
BiocManager::install("BSgenome.Hsapiens.UCSC.hg19")
BiocManager::install("TxDb.Hsapiens.UCSC.hg19.knownGene")
BiocManager::install("org.Hs.eg.db")

```

# load packages
```{r load packages}
library(Gviz) # 1.42.1
library(GenomicRanges)
library(BSgenome.Hsapiens.UCSC.hg19)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
```

## TEMPORARY loading bdg
```{r temp load bdg}
bgFile <- system.file("extdata/test.bedGraph", package = "Gviz")
dTrack.ATAC1 <- DataTrack(range = bgFile, genome = "hg19", type = "l", 
                     chromosome = "chr6", name = "bedGraph")
dTrack.ATAC2 <- dTrack.ATAC1
```

# Extract coordinates for a gene

Below I write a function that returns the chromosome and start/end coordinates for a given HGNC gene symbol. This will be useful when we assign coordinates to gviz for drawing. 
```{r extract coordinates for a given gene}
library(org.Hs.eg.db)

gene_coord <- function(symbol, Txdb = TxDb.Hsapiens.UCSC.hg19.knownGene){
  symbol <- toupper(symbol) # make gene symbol upper case
  
  # find the entrez id for that symbol
  entrez_id <- select(org.Hs.eg.db, keys=symbol, keytype="SYMBOL", columns="ENTREZID")
  
  # find the coordinates of that gene based on entrez id
  coords <- genes(TxDb.Hsapiens.UCSC.hg19.knownGene, filter = list(gene_id=entrez_id$ENTREZID))
  
  print(paste("The ENTREZID for the following gene:",symbol,"is:",entrez_id$ENTREZID))
  
  print(paste(
      "The coordinates for the following gene:",
      symbol,
      "is:",
      as.character(seqnames(coords)),start(coords),"-",end(coords)
  ))
  
}

# example function call
gene_coord("CA9")
```

# DataTrack construction
Next, we create DataTracks for each bigwig we want visualized using Gviz.

```{r data track from bam}
# this is an example from the Gviz vignette
chr = "chr1"
bamFile <- system.file("extdata/test.bam", package = "Gviz")

dTrack1 <- DataTrack(range = bamFile, genome = "hg19", type = "l", 
                     name = "Coverage", window = -1, 
                     chromosome = chr)

dTrack1
plotTracks(dTrack1, from = 189990000, to = 190000000)
```

## load bigwigs
Here we load the bigwig (bw) files generated from MACS2 from our ATAC-seq experiment.
```{r data track from bigwig}
# set chromosome
chr <- "chr1"

# path to bigwig
bwFile1 <-"J:/genome-viewer/heatmap/RW555-1_treat_pileup.bw"
bwFile2 <-"J:/genome-viewer/heatmap/RW555-2_treat_pileup.bw"

# create a DataTrack for each bigwig
dTrack.ATAC1 <- DataTrack(range = bwFile1, genome = "hg19", type = "l", 
                     name = "Coverage", window = -1, 
                     chromosome = chr)

dTrack.ATAC2 <- DataTrack(range = bwFile2, genome = "hg19", type = "l", 
                     name = "Coverage", window = -1, 
                     chromosome = chr)
# plot each DataTrack
plotTracks(dTrack.ATAC1, from = 189990000, to = 190000000)
plotTracks(dTrack.ATAC2, from = 189990000, to = 190000000)
```

## Plot DataTracks with additional tracks
Next, we add additional tracks to generate a genome view that is similar to what IGV or UCSC genome browser would provide.
```{r data track from bigwig chr6}
gen = "hg19"
chr = "chr6"

# add genome coordinates track
gtrack <- GenomeAxisTrack()

# add ideogram track (visual representation of chromosome)

itrack <- IdeogramTrack(genome = "hg19", chromosome = chr)

# add sequence track
strack <- SequenceTrack(Hsapiens, chromosome = chr)

# create a DataTrack for each bigwig
## add groups to allow additional params (ie color) to data tracks
dTrack.ATAC1 <- DataTrack(range = bwFile1, genome = "hg19", type = "l", 
                     name = "Signal", window = -1, 
                     chromosome = chr,
                     groups = factor("normoxia", levels =c("normoxia","hypoxia")),
                     legend = TRUE)

dTrack.ATAC2 <- DataTrack(range = bwFile2, genome = "hg19", type = "l", 
                     name = "Signal", window = -1, 
                     chromosome = chr)

# manually set display params (use this if you don't want to assign params during object instantiation)
displayPars(dTrack.ATAC2) <- list(groups = factor("hypoxia", 
              levels = c("normoxia","hypoxia")), legend = TRUE) 

# plot two atac signal tracks with additional tracks
plotTracks(list(gtrack,
                itrack,
                dTrack.ATAC1,
                dTrack.ATAC2,
                strack), 
           from = 35000000, to = 40000000)

# plot two atac signal tracks overlaid

ot1 <- OverlayTrack(trackList=list(dTrack.ATAC1, dTrack.ATAC2))

plotTracks(
  list(gtrack,
       itrack,
       ot1,
       strack),
  from = 35000000,
  to = 40000000,
  col = c("red","blue") # assign colors to the DataTrack groups defined earlier (ie. normoxia, hypoxia)
)


```

# Obtain GeneRegionTrack
Often it is important to include a GeneRegionTrack to give context to which gene is being visualized through gene models. Both TxDb and BiomaRt can be used to retrieve this information, however, the track info is not identical.

```{r GeneRegionTrack}
chr <- "chr6"
# TxDb
library(GenomicFeatures)
samplefile <- system.file("extdata", "hg19_knownGene_sample.sqlite", 
                          package = "GenomicFeatures")
txdb <- loadDb(samplefile)
GeneRegionTrack(txdb)

# biomart
library(biomaRt)
bm <- useEnsembl(host = "https://grch37.ensembl.org", 
              biomart = "ENSEMBL_MART_ENSEMBL", 
              dataset = "hsapiens_gene_ensembl")
```

```{r comparing TxDb vs biomaRt}
# plot TxDb
txTr <- GeneRegionTrack(txdb,
                        chromosome = chr,
                        start = 35000000,
                        end = 40000000)

plotTracks(txTr)
plotTracks(txTr, from = 35000000,
           to = 40000000)

# plot biomaRt
biomTrack <-
  BiomartGeneRegionTrack(
    genome = "hg19",
    chromosome = chr,
    start = 35000000,
    end = 40000000,
    name = "ENSEMBL",
    biomart = bm
  )
biomTrack_filt <-
  BiomartGeneRegionTrack(
    genome = "hg19",
    chromosome = chr,
    start = 35000000,
    end = 40000000,
    name = "ENSEMBL",
    filter = list(with_refseq_mrna=TRUE), # filtering to RefSeq models only
    biomart = bm
  )
biomTrack_filt_collapse <-
  BiomartGeneRegionTrack(
    genome = "hg19",
    chromosome = chr,
    start = 35000000,
    end = 40000000,
    name = "ENSEMBL",
    filter = list(with_refseq_mrna=TRUE), # filtering to RefSeq models only
    collapseTranscripts = "longest",
    biomart = bm
  )

plotTracks(list(biomTrack,biomTrack_filt, biomTrack_filt_collapse), from = 35000000, to = 40000000)



ot1 <- OverlayTrack(trackList=list(dTrack.ATAC1, dTrack.ATAC2))
# plot in the context of ATAC result
plotTracks(
  list(
       #gtrack,
       #itrack,
       ot1,
       txTr, biomTrack),
  from = 35000000,
  to = 40000000,
  col = c("red", "blue")
)
```

# Plot Gviz Genome Visualization for genes of interest
## CA9
```{r CA9}
# determine locus location
gene_coord("CA9")
# CA9 Locus (coordinates below are from IGV and include upstream; not exactly same coords as from TxDb)
#   chr9
#   from = 35671925, to = 35683156

chr <-"chr9"

# add genome coordinates track
gtrack <- GenomeAxisTrack()

# add ideogram track (visual representation of chromosome)

itrack <- IdeogramTrack(genome = "hg19", chromosome = chr)

# biomart GeneRegionTrack at CA9
## a) select CA9 locus via gene symbol
biomTrack <- BiomartGeneRegionTrack(genome = "hg19", name = "ENSEMBL", 
                                    symbol = "CA9", biomart = bm)

plotTracks(biomTrack, 
           transcriptAnnotation = "symbol")
## b) select CA9 locus via ensembl id
biomTrack1 <- BiomartGeneRegionTrack(genome = "hg19", name = "ENSEMBL", 
                                    gene = "ENSG00000107159", biomart = bm,
                                    col = "transparent")
plotTracks(biomTrack1, 
           transcriptAnnotation = "symbol")


# add additional params (ie color) to data tracks
dTrack.ATAC1 <- DataTrack(range = bwFile1, genome = "hg19", type = "l", 
                     name = "Signal", window = -1, 
                     chromosome = chr,
                     groups = factor("normoxia", levels =c("normoxia","hypoxia")),
                     legend = TRUE)

dTrack.ATAC2 <- DataTrack(range = bwFile2, genome = "hg19", type = "l", 
                     name = "Signal", window = -1, 
                     chromosome = chr)
displayPars(dTrack.ATAC2) <- list(groups = factor("hypoxia", 
              levels = c("normoxia","hypoxia")), legend = TRUE) # manually set display params 

ot1 <- OverlayTrack(trackList=list(dTrack.ATAC1, dTrack.ATAC2))

# ######
# plot #
########

# stacked version
plotTracks(
  list(itrack,
       gtrack,
       dTrack.ATAC1,
       dTrack.ATAC2,
       biomTrack1),
  from = 35671925,
  to = 35683156,
  col = c("red", "blue"),
  transcriptAnnotation = "symbol",
  # set group colors
  ylim = c(0, 3) # The range of the y-axis scale
)

# overlay version
plotTracks(
  list(itrack,
       gtrack,
       ot1,
       biomTrack1),
  from = 35671925,
  to = 35683156,
  col = c("red", "blue"),
  # set group colors
  transcriptAnnotation = "symbol",
  ylim = c(0, 3) # The range of the y-axis scale
)

```

# Dynamic Y-axis scaling
Gviz can overlay several DataTracks using the `OverlayTrack()` function, but the y-axis range is controlled by the first DataTrack. One must explicitly specify the y-axis range that is shared across all those overlayed tracks for proper comparison. The default behavior is to take the y-axis of the first DataTrack within the list of DataTracks being overlayed. This can be problematic if other DataTracks have different maximal signals, which results in a suboptimal y-axis range being set.

Here I write a function that can dynamically scale the Y-axis given one or more bigwigs that are plotted together.
## one sample
```{r dyanamic y-axis scaling function}
# Function to get max signal from bigWig in a given range
getMaxSignal <- function(bigwig_file, chr, start, end) {
  # Import specific genomic range from bigWig
  bw <- import(bigwig_file, 
              format = "bigWig",
              which = GRanges(seqnames = chr,
                             ranges = IRanges(start = start, 
                                            end = end)))
  
  # Return maximum signal value
  return(max(bw$score))
}

# Create and plot track with dynamic y-axis
createScaledTrack <- function(bigwig_file, chr, start, end, 
                             track_name = "Signal Track") {
  # Get maximum signal for scaling
  max_signal <- getMaxSignal(bigwig_file, chr, start, end)
  
  print(paste("The max signal for this bigwig is", max_signal))
  # Create DataTrack with adjusted y-limits
  dt <- DataTrack(range = bigwig_file,
                  chromosome = chr,
                  from = start,
                  to = end,
                  name = track_name,
                  type = "l",
                  ylim = c(0, max_signal * 1.1)) # Add 10% padding
  
  return(dt)
}
```

```{r testing dynamic y-axis scaling}
# Example usage
## comment out depending on PC or macbook:
# bigwig_file <- "data/ENCODE/ENCFF923TGK.bigWig"
bigwig_file <- bwFile1

chr <- "chr9"
start <- 35671925
end <- 35683156


# Create track with automatic scaling
scaled_track <- createScaledTrack(bigwig_file, chr, start, end)

# Plot track
plotTracks(scaled_track, from = start, to = end)

# Combine with other tracks:
genome_axis <- GenomeAxisTrack()
plotTracks(list(genome_axis, scaled_track), from = start, to = end)

# compare unscaled to scaled
unscaled_track <- DataTrack(range = bigwig_file,
                  chromosome = chr,
                  from = start,
                  to = end,
                  name = "unscaled",
                  type = "l"
                  ) 

plotTracks(
  list(itrack,
       gtrack,
       unscaled_track,
       scaled_track,
       biomTrack1),
  from = 35671925,
  to = 35683156,
  col = c("red", "blue"),
  # set group colors
  transcriptAnnotation = "symbol"
)
```
Typically, more than one bigwig file is used to construct a figure. For example, if we are comparing control vs treatment or in the case below, normoxia vs hypoxia.
Below we adjust the function to take in more than one input bigwig file:
## one or more samples
```{r one or more bw y-axis scaling}
# Function to get max signal from one or more bigWig files in a given range
getMaxSignal <- function(bigwig_files, chr, start, end) {
  # Handle single file input
  if (is.character(bigwig_files) && length(bigwig_files) == 1) {
    bigwig_files <- list(bigwig_files)
  }
  
  # Get maximum signal across all files
  max_signals <- sapply(bigwig_files, function(bw) {
    # Import specific genomic range from bigWig
    data <- import(bw,
                   format = "bigWig",
                   which = GRanges(
                     seqnames = chr,
                     ranges = IRanges(start = start,
                                      end = end)
                   ))
    
    if (length(data) == 0)
      return(0)
    return(max(data$score))
  })
  
  return(max(max_signals))
}

# Create and plot tracks with dynamic y-axis
createScaledTracks <- function(bigwig_files,
                               chr,
                               start,
                               end,
                               track_names = NULL,
                               colors = NULL,
                               padding_factor = 1.1,
                               overlay = "no") {
  # Handle input parameters
  if (is.character(bigwig_files) && length(bigwig_files) == 1) {
    bigwig_files <- list(bigwig_files)
  }
  
  # Generate default track names if not provided
  if (is.null(track_names)) {
    track_names <- paste("Signal", seq_along(bigwig_files))
  } else if (length(track_names) != length(bigwig_files)) {
    stop("Number of track names must match number of bigWig files")
  }
  
  # Generate default colors if not provided
  if (is.null(colors)) {
    colors <- rainbow(length(bigwig_files))
  } else if (length(colors) != length(bigwig_files)) {
    stop("Number of colors must match number of bigWig files")
  }
  
  # Sample Grouping
  ## INSERT sample grouping code
  
  # Get maximum signal across all files
  max_signal <- getMaxSignal(bigwig_files, chr, start, end)
  
  
  # tracks <- mapply(function(bw, name, col) {
  #   DataTrack(range = bw,
  #             chromosome = chr,
  #             from = start,
  #             to = end,
  #             name = name,
  #             type = "l",
  #             col = col,
  #             #groups = factor(track_names[1],
  #             groups = factor(name,
  #             levels = track_names),
  #             legend = TRUE,
  #             ylim = c(0, max_signal * padding_factor))
  #
  # }, bigwig_files, track_names, colors, SIMPLIFY = FALSE)
  
  # Create DataTrack for each bigWig file
  tracks <- list()
  for (i in seq_along(bigwig_files)) {
    tracks[[i]] <- DataTrack(
      range = bigwig_files[[i]],
      chromosome = chr,
      from = start,
      to = end,
      name = track_names[i],
      type = "l",
      col = colors[i],
      # Explicitly assign track color by index
      baseline = 0,
      # Ensure histogram starts at 0
      col.baseline = "black",
      # set baseline color
      groups = factor(track_names[i],
                      levels = track_names),
      legend = TRUE,
      ylim = c(0, max_signal * padding_factor)
    )
  }
  
  # return scaled tracks:
  # option to overlay tracks
  if (tolower(overlay) == "yes") {
    tracks[[1]]@name <- "signal" # set shared y-axis title
    overlayTrack <- OverlayTrack(trackList = tracks)
    return(overlayTrack)
  } else{
    # default to stack tracks
    return(tracks)
  }
  
}

```

```{r example use of one or more samples}
# Example usage:

# Single file
single_bigwig <- "J:/genome-viewer/heatmap/RW555-1_treat_pileup.bw"
single_track <- createScaledTracks(single_bigwig, "chr9", 35671925, 35683156)

# Multiple files
bigwig_files <- list(
  "J:/genome-viewer/heatmap/RW555-1_treat_pileup.bw",
  "J:/genome-viewer/heatmap/RW555-2_treat_pileup.bw"
)

# Custom track names and colors
track_names <- c("normoxia", "hypoxia")
track_colors <- c("red", "blue")

# Create tracks with custom parameters
chr = "chr9"
start = 35671925
end = 35683156

scaledTracks <- createScaledTracks(
  bigwig_files = bigwig_files,
  chr = chr,
  start = start,
  end = end,
  track_names = track_names,
  colors = track_colors,
  padding_factor = 1.2  # 20% padding above max signal
)

# Plot tracks with genome axis
gtrack <- GenomeAxisTrack()
itrack <- IdeogramTrack(genome = "hg19", chromosome = chr)

plotTracks(c(list(itrack,
                  genome_axis), 
                  scaledTracks),
           from = start, to = end)

plotTracks(c(list(itrack,
                  gtrack),
             scaledTracks,
             biomTrack1),
           from = start,
           to = end,
           biomTrack1,
           col = c("red", "blue"),
           transcriptAnnotation = "symbol")

# test overlay
scaledTracks2 <- createScaledTracks(
  bigwig_files = bigwig_files,
  chr = chr,
  start = start,
  end = end,
  track_names = track_names,
  colors = track_colors,
  padding_factor = 1.2,  # 20% padding above max signal
  overlay = "yes"
)

plotTracks(
  c(list(itrack,
         gtrack),
    scaledTracks2,
    biomTrack1),
  from = start,
  to = end,
  background.title = "transparent",
  fontcolor.title = "black",
  # y-axis colors
  col.axis = "black",
  # y-axis colors
  col = c("red", "blue"),
  transcriptAnnotation = "symbol"
)


# col = c("red", "blue"),
  # set group colors
  #ylim = c(0, 3) # The range of the y-axis scale
```
