---
title: "gvis_tutorial"
output: html_notebook
---
#
# Introduction

Below is a slightly modified copy of the Gviz [vignette](https://bioconductor.org/packages/devel/bioc/vignettes/Gviz/inst/doc/Gviz.html) with minor additional comments and modifications.

# install packages
```{r installation, eval = FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("Gviz")
BiocManager::install("BSgenome.Hsapiens.UCSC.hg19")
```

# load packages
```{r load packages}
library(Gviz) # 1.42.1
library(GenomicRanges)
library(BSgenome.Hsapiens.UCSC.hg19)
```

# Basic Features: CpG island example
```{r}
data(cpgIslands)
class(cpgIslands)

chr <- as.character(unique(seqnames(cpgIslands)))
gen <- genome(cpgIslands)
atrack <- AnnotationTrack(cpgIslands, name = "CpG")
```

## plots cpg island track
```{r plot cpg island track}
# plot cpg island track alone
plotTracks(atrack)

# add genome coordinates track
gtrack <- GenomeAxisTrack()

# plot cpg with genome coordinates

plotTracks(list(gtrack, atrack))

# add ideogram track (visual representation of chromosome)

itrack <- IdeogramTrack(genome = gen, chromosome = chr)

# plot tracks (cpg islands, genome coordinates, and chr7 ideogram)

plotTracks(list(itrack,
                gtrack,
                atrack))
```
## Using gene models

```{r gene models}
data(geneModels)

# use GeneRegionTrack
grtrack <- GeneRegionTrack(geneModels, genome = gen,
                           chromosome = chr, name = "Gene Model")
plotTracks(list(itrack, gtrack, atrack, grtrack))
```
## Zooming in
The default plotting view automatically shows the leftmost items to the right most item. Here we specify a range to plot.

```{r zoom}
# Specify range to plot
plotTracks(list(itrack, gtrack, atrack, grtrack),
           from = 26700000, to = 26750000)

# Controls the zoom state with extend.left and extend.right 
plotTracks(list(itrack, gtrack, atrack, grtrack),
           extend.left = 0.5, extend.right = 1000000)

# drop bounding borders of exons
plotTracks(list(itrack, gtrack, atrack, grtrack), 
           extend.left = 0.5, extend.right = 1000000, col = NULL)
```
## Add genomic sequence track
If we zoom in to the base-sequence level, we can add a sequence track provided from the package `BSgenome.Hsapiens.UCSC.hg19`. Note, other `BSgenome` packages are available depending on the genome of interest.

```{r genome sequence track}
# full hg19 genome sequence
BSgenome.Hsapiens.UCSC.hg19::Hsapiens

# add genome sequence track
strack <- SequenceTrack(Hsapiens, chromosome = chr)

# plot
plotTracks(list(itrack, gtrack, atrack, grtrack, strack), 
           from = 26591822, to = 26591852, cex = 0.8)

```
## Data Track for numeric data visualization
So far the various tracks provide a solid backbone for genomic visualization. However, we want to add additional tracks from experimental data for example from ChIP-seq. This is where the DataTrack comes in and is the meat and potatoes of Gviz, allowing several types of visualization:
1. histograms
2. dot plots
3. box and whisker plots

```{r DataTrack}

# randomly sampled data: dot plot (default)
set.seed(255)
lim <- c(26700000, 26750000)
coords <- sort(c(lim[1], 
                 sample(seq(from = lim[1], to = lim[2]), 99), 
                 lim[2]))
dat <- runif(100, min = -10, max = 10)
dtrack <- DataTrack(data = dat, start = coords[-length(coords)],
                    end = coords[-1], chromosome = chr, genome = gen, 
                    name = "Uniform")

# plot dot plot data track
plotTracks(list(itrack, gtrack, atrack, grtrack, dtrack), 
           from = lim[1], to = lim[2])

# plot histogram data track
plotTracks(list(itrack, gtrack, atrack, grtrack, dtrack), 
           from = lim[1], to = lim[2],
           type = "histogram")
```

# Advanced Features
## Display parameters
```{r display}
grtrack <- GeneRegionTrack(geneModels, genome = gen, chromosome = chr, 
                           name = "Gene Model",
                           transcriptAnnotation = "symbol",
                           background.title = "brown")

# show display parameters
head(displayPars(grtrack))

# modify display parameters directly in a given track

displayPars(grtrack) <- list(background.panel = "#FFFEDB", col = NULL)
head(displayPars(grtrack))

# plot
plotTracks(list(
  itrack,
  gtrack,
  atrack,
  grtrack
))

# apply display parameters to all tracks directly via plotTracks()
## some track objects will ignore these params if they are not compatible
plotTracks(
  list(itrack,
       gtrack,
       atrack,
       grtrack),
  background.panel = "#FFFEDB",
  background.title = "darkblue"
)
```
### show all possible display params for a given track object
```{r show params}
dp <- availableDisplayPars(grtrack)
tail(dp)
```

## Schemes
Schemes are used for setting global display parameters in a central location. Within a scheme are nested named lists corresponding to the various track class named and their associated display parameters.

```{r setting schemes}
getOption("Gviz.scheme") ## [1] "default"

# make a new scheme called "myScheme"
scheme <- getScheme()
scheme$GeneRegionTrack$fill <- "salmon"
scheme$GeneRegionTrack$col <- NULL
scheme$GeneRegionTrack$transcriptAnnotation <- "transcript"
addScheme(scheme, "myScheme")

# apply myScheme
options(Gviz.scheme = "myScheme")
grtrack <- GeneRegionTrack(geneModels, genome = gen,
                           chromosome = chr, name = "Gene Model")
# plot
plotTracks(grtrack)

# reset scheme to default
options(Gviz.scheme="default")
grtrack <- GeneRegionTrack(geneModels, genome = gen, chromosome = chr,
                           name = "Gene Model",
                           transcriptAnnotation = "symbol")
```

# Track Classes

## GenomeAxisTrack
```{r GenomeAxisTrack}
# highlight regions on GenomeAxisTrack
axisTrack <- GenomeAxisTrack(range=IRanges(start = c(2e6, 4e6),
                                           end = c(3e6, 7e6),
                                           names = rep("N-stretch", 2)))
plotTracks(axisTrack, from = 1e6, to = 9e6)
plotTracks(axisTrack, from = 1e6, to = 9e6, showId = TRUE) # names supplied to range argument is shown

# scale is used to represent the fraction of the plotting region
plotTracks(axisTrack, from = 1e6, to = 9e6, scale = 0.5, 
           labelPos = "below")
plotTracks(axisTrack, from = 1e6, to = 9e6, scale = 1, 
           labelPos = "below")

```
## IdeogramTrack
```{r ideogram}

ideoTrack <- IdeogramTrack(genome = "hg19", chromosome = "chrX")
plotTracks(ideoTrack, from = 85e6, to = 129e6)

# hide track name
plotTracks(ideoTrack, from = 85e6, to = 129e6, showId = FALSE)

# add chromosome band labels
plotTracks(ideoTrack, from = 85e6, to = 129e6, showId = FALSE, 
           showBandId = TRUE, cex.bands = 0.5)
# change centromere shape
plotTracks(ideoTrack, from = 85e6, to = 129e6, showId = FALSE,
           centromereShape = "circle")
```
## DataTrack

```{r datatrack}
data(twoGroups) # granges object

# load granges data to DataTrack
dTrack <- DataTrack(twoGroups, name = "uniform")

# plot
plotTracks(dTrack) # dotplot (default)

plotTracks(dTrack, type = "l") # lines plot

plotTracks(dTrack, type = "b") # dot and lines plot

plotTracks(dTrack, type = "a") # lines plot of average values

plotTracks(dTrack, type = "s") # stair steps (horizontal first)

plotTracks(dTrack, type = "S") # stair steps (vertical first)

plotTracks(dTrack, type = "histogram") # histogram
plotTracks(dTrack, type = "mountain") # ‘mountain-type’ plot relative to a baseline
plotTracks(dTrack, type = "polygon") # ‘polygon-type’ plot relative to a baseline
plotTracks(dTrack, type = "boxplot") # box and whisker plot
plotTracks(dTrack, type = "gradient") # false color image of the summarized values
plotTracks(dTrack, type = "heatmap") # false color image of the individual values
plotTracks(dTrack, type = "horizon") # Horizon plot indicating magnitude and direction of a change relative to a baseline
```
```{r datatrack combinations}
# overlay different types in the datatrack:
## boxplot, line plot of average, and grid lines
plotTracks(dTrack, type = c("boxplot", "a", "g"))
```
```{r heatmap plotting}
colnames(mcols(twoGroups))

plotTracks(dTrack, type = c("heatmap"), showSampleNames = FALSE, 
           cex.sampleNames = 0.6)

# showSampleNames corresponds to the column names
plotTracks(dTrack, type = c("heatmap"), showSampleNames = TRUE, 
           cex.sampleNames = 0.6)
```
Here the column names of the granges object can be added as labels.

### Data grouping
Here groups parameter is used to group the individual samples in the granges object.

```{r data grouping}

# grouped lines plot
plotTracks(dTrack,
           groups = rep(c("control", "treated"), each = 3),
           type = c("a", "p", "confint"))

# turn off legend
plotTracks(
  dTrack,
  groups = rep(c("control", "treated"), each = 3),
  type = c("a", "p", "confint"),
  legend = FALSE
)

# grouped horizon plot
data(dtHoriz)
dtHoriz <- dtHoriz[1:6,]
plotTracks(
  dtHoriz,
  type = "horiz",
  groups = rownames(values(dtHoriz)),
  showSampleNames = TRUE,
  cex.sampleNames = 0.6,
  separator = 1
)
```

### Building DataTrack Object from files
Here the range argument is not a GRanges object as previous examples above, but an externally loaded file, such as a wig, bigwig, or bedgraph.

Below, an entire bedgraph file, which is not indexed, is loaded.

```{r datatrack from bedgraph}
bgFile <- system.file("extdata/test.bedGraph", package = "Gviz")
dTrack2 <- DataTrack(range = bgFile, genome = "hg19", type = "l", 
                     chromosome = "chr19", name = "bedGraph")
class(dTrack2)
plotTracks(dTrack2)
```

```{r data track from bam}
bamFile <- system.file("extdata/test.bam", package = "Gviz")
dTrack4 <- DataTrack(range = bamFile, genome = "hg19", type = "l", 
                     name = "Coverage", window = -1, 
                     chromosome = "chr1")
class(dTrack4)

dTrack4
plotTracks(dTrack4, from = 189990000, to = 190000000)
```
Here an indexed bam file is streamed for plotting from a specific range. This reduces the memory footprint as real bam files can be quite large.

### Data Transformations

Below is a synthetic example to demonstrate transformations we can do on the fly.

```{r data transformation}
dat <- sin(seq(pi, 10*pi, len=500))
dTrack.big <- DataTrack(start = seq(1, 100000, len = 500), width = 15, 
                        chromosome = "chrX", genome = "hg19", 
                        name = "sinus",
                        data = sin(seq(pi, 5 * pi, len = 500)) * 
                          runif(500, 0.5, 1.5))
plotTracks(dTrack.big, type="hist")

# bin the data
# using window and aggregation params
plotTracks(dTrack.big, type = "hist", window = 50)

# smoothing using running window
plotTracks(dTrack.big, type = "hist", window = -1, windowSize = 2500)

# transformation using a function
## plot values greater than zero
plotTracks(dTrack.big, type = "l", 
           transformation = function(x) { x[x < 0] <- 0; x })

```

```{r aggregateGroups}
# without aggregate groups
plotTracks(
  dTrack,
  groups = rep(c("control", "treated"), each = 3),
  type = c("b"),
  aggregateGroups = FALSE
)

# aggregateGroups
## display average value for each group and each position
## aggregation is mean by default
plotTracks(
  dTrack,
  groups = rep(c("control", "treated"), each = 3),
  type = c("b"),
  aggregateGroups = TRUE
)

# display the maximum group values instead of mean
plotTracks(
  dTrack,
  groups = rep(c("control", "treated"), each = 3),
  type = c("b"),
  aggregateGroups = TRUE,
  aggregation = "max"
)

# adjust color
plotTracks(
  dTrack,
  groups = rep(c("control", "treated"), each = 3),
  type = c("b"),
  aggregateGroups = TRUE,
  aggregation = "max",
  col = c("red","green")
)
```
```

## AnnotationTrack
