---
title: "gvis_atac"
output: html_notebook
---

# Introduction

# install packages

Here we install Gviz and their dependencies to allow for genome visualization. Additional packages such as `BSgenome` and `TxDb` contain sequence information and genomic coordinate info respectively. Specifically, we install the hg19 version of these packages, since we know we will be using hg19 aligned ATAC-seq track and signal files. Typically hg38 is preferred for newer data analysis, but it is important to note the genomic coordinates will differ between hg19 and hg38.

```{r installation, eval = FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("Gviz")
BiocManager::install("BSgenome.Hsapiens.UCSC.hg19")
BiocManager::install("TxDb.Hsapiens.UCSC.hg19.knownGene")
BiocManager::install("org.Hs.eg.db")

```

# load packages

```{r load packages}
library(Gviz) # 1.42.1
library(rtracklayer)
library(GenomicRanges)
library(BSgenome.Hsapiens.UCSC.hg19)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
```

# Prefetch UCSC data for offline use

In order to create ideogram tracks on the computing cluster that has no internet access, we must prefetch the data from UCSC and save it for later use. This will allow us to bypass the fetching step when we have no internet on the cluster.

```{r prefetch ucsc data, eval = FALSE}
# Download cytoBand data directly for hg38
download.file(
  "http://hgdownload.cse.ucsc.edu/goldenPath/hg38/database/cytoBand.txt.gz",
  destfile = "cytoBand_hg38.txt.gz"
)

# Download cytoBand data directly for hg19
download.file(
  "http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/cytoBand.txt.gz",
  destfile = "cytoBand_hg19.txt.gz"
)

# Read the downloaded file
cytoBands <- read.table(
  "cytoBand_hg19.txt.gz",
  sep = "\t",
  col.names = c("chrom", "chromStart", "chromEnd", "name", "gieStain")
)

# Create ideogram track
ideoTrack <- IdeogramTrack(genome="hg38", bands=cytoBands, chromosome="chr1")

# Plot
plotTracks(list(ideoTrack), from = 189990000, to = 190000000)

# save bands for offline use
saveRDS(cytoBands, file="hg19_cytobands.rds")

# load bands for offline use
cytoBands <- readRDS("hg19_cytobands.rds")
```

## TEMPORARY loading bdg
Here we can load a bedgraph file included in the Gviz package for testing.
```{r temp load bdg}
bgFile <- system.file("extdata/test.bedGraph", package = "Gviz")
dTrack.ATAC1 <- DataTrack(range = bgFile, genome = "hg19", type = "l", 
                     chromosome = "chr6", name = "bedGraph")
dTrack.ATAC2 <- dTrack.ATAC1
```

# Extract coordinates for a gene

Below I write a function that returns the chromosome and start/end coordinates for a given HGNC gene symbol. This will be useful when we assign coordinates to gviz for drawing.

```{r extract coordinates for a given gene}
library(org.Hs.eg.db)

#' Find Genome coordinates for a Gene
#'
#' @param symbol A character representing the gene symbol 
#' @param Txdb A GenomicFeatures object. The TxDb specific to your genome build (hg19). 
#'
#' @return Returns a Granges object. The Granges object will return the relevant seqnames, #' ranges, strand, and gene_id (ENTREZID) specific to the symbol provided to the function.
#' @export
#'
#' @examples gene_coord("CA9")
#' 
gene_coord <- function(symbol, Txdb = TxDb.Hsapiens.UCSC.hg19.knownGene){
  symbol <- toupper(symbol) # make gene symbol upper case
  
  # find the entrez id for that symbol
  entrez_id <- select(org.Hs.eg.db, keys=symbol, keytype="SYMBOL", columns="ENTREZID")
  
  # find the coordinates of that gene based on entrez id
  coords <- genes(TxDb.Hsapiens.UCSC.hg19.knownGene, filter = list(gene_id=entrez_id$ENTREZID))
  
  print(paste("The ENTREZID for the following gene:",symbol,"is:",entrez_id$ENTREZID))
  
  print(paste(
      "The coordinates for the following gene:",
      symbol,
      "is:",
      as.character(seqnames(coords)),start(coords),"-",end(coords)
  ))
class(coords)  
return(coords)
}

# example function call
gene_coord("CA9")
```

# DataTrack construction

Next, we create DataTracks for each bigwig we want visualized using Gviz.

```{r data track from bam}
# this is an example from the Gviz vignette
chr = "chr1"
bamFile <- system.file("extdata/test.bam", package = "Gviz")

dTrack1 <- DataTrack(range = bamFile, genome = "hg19", type = "l", 
                     name = "Coverage", window = -1, 
                     chromosome = chr)

dTrack1
plotTracks(dTrack1, from = 189990000, to = 190000000)
```

## load bigwigs

Here we load the bigwig (bw) files generated from MACS2 from our ATAC-seq experiment.

```{r data track from bigwig}
# set chromosome
chr <- "chr1"

# path to bigwig
bwFile1 <-"J:/genome-viewer/heatmap/RW555-1_treat_pileup.bw"
bwFile2 <-"J:/genome-viewer/heatmap/RW555-2_treat_pileup.bw"

# create a DataTrack for each bigwig
dTrack.ATAC1 <- DataTrack(range = bwFile1, genome = "hg19", type = "l", 
                     name = "Coverage", window = -1, 
                     chromosome = chr)

dTrack.ATAC2 <- DataTrack(range = bwFile2, genome = "hg19", type = "l", 
                     name = "Coverage", window = -1, 
                     chromosome = chr)
# plot each DataTrack
plotTracks(dTrack.ATAC1, from = 189990000, to = 190000000)
plotTracks(dTrack.ATAC2, from = 189990000, to = 190000000)
```

## Plot DataTracks with additional tracks

Next, we add additional tracks to generate a genome view that is similar to what IGV or UCSC genome browser would provide.

```{r data track from bigwig chr6}
gen = "hg19"
chr = "chr6"

# add genome coordinates track
gtrack <- GenomeAxisTrack()

# add ideogram track (visual representation of chromosome)

itrack <- IdeogramTrack(genome = "hg19", chromosome = chr)

# add sequence track
strack <- SequenceTrack(Hsapiens, chromosome = chr)

# create a DataTrack for each bigwig
## add groups to allow additional params (ie color) to data tracks
dTrack.ATAC1 <- DataTrack(range = bwFile1, genome = "hg19", type = "l", 
                     name = "Signal", window = -1, 
                     chromosome = chr,
                     groups = factor("normoxia", levels =c("normoxia","hypoxia")),
                     legend = TRUE)

dTrack.ATAC2 <- DataTrack(range = bwFile2, genome = "hg19", type = "l", 
                     name = "Signal", window = -1, 
                     chromosome = chr)

# manually set display params (use this if you don't want to assign params during object instantiation)
displayPars(dTrack.ATAC2) <- list(groups = factor("hypoxia", 
              levels = c("normoxia","hypoxia")), legend = TRUE) 

# plot two atac signal tracks with additional tracks
plotTracks(list(gtrack,
                itrack,
                dTrack.ATAC1,
                dTrack.ATAC2,
                strack), 
           from = 35000000, to = 40000000)

# plot two atac signal tracks overlaid

ot1 <- OverlayTrack(trackList=list(dTrack.ATAC1, dTrack.ATAC2))

plotTracks(
  list(gtrack,
       itrack,
       ot1,
       strack),
  from = 35000000,
  to = 40000000,
  col = c("red","blue") # assign colors to the DataTrack groups defined earlier (ie. normoxia, hypoxia)
)


```

# Obtain GeneRegionTrack

Often it is important to include a GeneRegionTrack to give context to which gene is being visualized through gene models. Both TxDb and BiomaRt can be used to retrieve this information, however, the track info is not identical.

TxDb may also be preferable if internet access is not available such as operating on a retricted node on a computing cluster. Despite prefeching the mart object from biomaRt, it still tries to fetch from ensembl.org.

```{r TxDb setup}
# TxDb
library(GenomicFeatures)
samplefile <- system.file("extdata", "hg19_knownGene_sample.sqlite", 
                          package = "GenomicFeatures")
txdb <- loadDb(samplefile)
txdb_hg19 <- TxDb.Hsapiens.UCSC.hg19.knownGene
```

```{r GeneRegionTrack}
chr <- "chr6"

GeneRegionTrack(txdb)

# biomart
library(biomaRt)
bm <- useEnsembl(host = "https://grch37.ensembl.org", 
              biomart = "ENSEMBL_MART_ENSEMBL", 
              dataset = "hsapiens_gene_ensembl")
```

```{r comparing TxDb vs biomaRt}
# make GeneRegionTrack with TxDb
txTr <- GeneRegionTrack(txdb,
                        chromosome = chr,
                        start = 35000000,
                        end = 40000000)
txTr2 <- GeneRegionTrack(txdb_hg19,
                        chromosome = chr,
                        start = 35000000,
                        end = 40000000)

# Map transcript IDs to gene symbols (TxDb doesn't store gene symbols)
symbols <- unlist(mapIds(org.Hs.eg.db, gene(txTr2), "SYMBOL", "ENTREZID", multiVals = "first"))
symbol(txTr2) <- symbols[gene(txTr2)]


# plot GeneRegionTracks made with Txdb
#plotTracks(txTr)
plotTracks(txTr, from = 35000000,
           to = 40000000)
plotTracks(list(txTr,txTr2), from = 35000000,
           to = 40000000)

plotTracks(txTr2, collapseTranscripts = "longest",
           transcriptAnnotation = "symbol")

#' Make GeneRegionTrack using TxDb and with proper Gene Symbol display
#'
#' @param txdb txdb object. with genome annotation appropriate for your genome ie. hg19
#' @param chr character. character that shows chromosome for gene of interest ie. chr7
#' @param start numeric. start coordinate for the gene
#' @param end numeric. end coordinate for the gene
#' @param name character. Label name for y-axis
#'
#' @return
#' @export
#'
#' @examples
MakeGeneRegionTrack_txdb <- function(txdb, chr, start, end, name = "gene"){
  gr <- GeneRegionTrack(txdb,
                        chromosome = chr,
                        start = start,
                        end = end,
                        name = name)
  # Map transcript IDs to gene symbols (TxDb doesn't store gene symbols)
  symbols <-
    unlist(mapIds(org.Hs.eg.db, gene(gr), "SYMBOL", "ENTREZID", multiVals = "first"))
  symbol(gr) <- symbols[gene(gr)]
  return(gr)
}

```

```{r comparing TxDb vs biomaRt2}
# plot biomaRt
biomTrack <-
  BiomartGeneRegionTrack(
    genome = "hg19",
    chromosome = chr,
    start = 35000000,
    end = 40000000,
    name = "ENSEMBL",
    biomart = bm
  )
biomTrack_filt <-
  BiomartGeneRegionTrack(
    genome = "hg19",
    chromosome = chr,
    start = 35000000,
    end = 40000000,
    name = "ENSEMBL",
    filter = list(with_refseq_mrna=TRUE), # filtering to RefSeq models only
    biomart = bm
  )
biomTrack_filt_collapse <-
  BiomartGeneRegionTrack(
    genome = "hg19",
    chromosome = chr,
    start = 35000000,
    end = 40000000,
    name = "ENSEMBL",
    filter = list(with_refseq_mrna=TRUE), # filtering to RefSeq models only
    collapseTranscripts = "longest",
    biomart = bm
  )

plotTracks(list(biomTrack,biomTrack_filt, biomTrack_filt_collapse), from = 35000000, to = 40000000)



ot1 <- OverlayTrack(trackList=list(dTrack.ATAC1, dTrack.ATAC2))
# plot in the context of ATAC result
plotTracks(
  list(
       #gtrack,
       #itrack,
       ot1,
       txTr, biomTrack),
  from = 35000000,
  to = 40000000,
  col = c("red", "blue")
)
```

# Plot Gviz Genome Visualization for genes of interest

## CA9

```{r CA9}
# determine locus location using our custom function gene_coord
coords <- gene_coord("CA9")

# assign locus info
chr <- as.character(seqnames(coords))
start <- start(coords)
end <- end(coords)

# CA9 Locus (coordinates below are from IGV and include upstream; not exactly same coords as from TxDb)
#   chr9
#   from = 35671925, to = 35683156


# add genome coordinates track
gtrack <- GenomeAxisTrack()

# add ideogram track (visual representation of chromosome)

itrack <- IdeogramTrack(genome = "hg19", chromosome = chr)

# biomart GeneRegionTrack at CA9
## a) select CA9 locus via gene symbol
biomTrack <- BiomartGeneRegionTrack(genome = "hg19", name = "ENSEMBL", 
                                    symbol = "CA9", biomart = bm)
biomTrack <-
  BiomartGeneRegionTrack(
    genome = "hg19",
    name = "ENSEMBL",
    symbol = "CA9",
    biomart = bm,
    col = "transparent",
    filter = list(with_refseq_mrna = TRUE),
    # filtering to RefSeq models only
    collapseTranscripts = "longest"
  )

plotTracks(biomTrack, 
           transcriptAnnotation = "symbol")
## b) select CA9 locus via ensembl id
biomTrack1 <- BiomartGeneRegionTrack(genome = "hg19", name = "ENSEMBL", 
                                    gene = "ENSG00000107159", biomart = bm,
                                    col = "transparent")
## c) select CA9 locus with ensembl id /w refseq models only and collapse longest
biomTrack2 <- BiomartGeneRegionTrack(
  genome = "hg19",
  name = "ENSEMBL",
  gene = "ENSG00000107159",
  biomart = bm,
  col = "transparent",
  filter = list(with_refseq_mrna = TRUE),
  # filtering to RefSeq models only
  collapseTranscripts = "longest"
)

plotTracks(list(biomTrack2,biomTrack1,biomTrack),
           transcriptAnnotation = "symbol")


# add additional params (ie color) to data tracks
dTrack.ATAC1 <- DataTrack(range = bwFile1, genome = "hg19", type = "l", 
                     name = "Signal", window = -1, 
                     chromosome = chr,
                     groups = factor("normoxia", levels =c("normoxia","hypoxia")),
                     legend = TRUE)

dTrack.ATAC2 <- DataTrack(range = bwFile2, genome = "hg19", type = "l", 
                     name = "Signal", window = -1, 
                     chromosome = chr)
displayPars(dTrack.ATAC2) <- list(groups = factor("hypoxia", 
              levels = c("normoxia","hypoxia")), legend = TRUE) # manually set display params 

ot1 <- OverlayTrack(trackList=list(dTrack.ATAC1, dTrack.ATAC2))

# GeneRegionTrack txdb with UCSC transcript ids
txTr3 <- GeneRegionTrack(txdb_hg19,
                        chromosome = as.character(seqnames(coords)),
                        start = start(coords),
                        end = end(coords))
# GeneRegionTrack using txdb with proper gene symbol display
txTr4 <- MakeGeneRegionTrack_txdb(txdb_hg19,
                        chr = as.character(seqnames(coords)),
                        start = start(coords),
                        end = end(coords))

# ######
# plot #
########

# stacked version
plotTracks(
  list(itrack,
       gtrack,
       dTrack.ATAC1,
       dTrack.ATAC2,
       biomTrack1),
  from = 35671925,
  to = 35683156,
  col = c("red", "blue"),
  transcriptAnnotation = "symbol",
  # set group colors
  ylim = c(0, 3) # The range of the y-axis scale
)

# overlay version
plotTracks(
  list(itrack,
       gtrack,
       ot1,
       biomTrack1,
       txTr3,
       txTr4),
  from = 35671925,
  to = 35683156,
  col = c("red", "blue"),
  # set group colors
  transcriptAnnotation = "symbol",
  ylim = c(0, 3) # The range of the y-axis scale
)


plotTracks(
  list(itrack,
       gtrack,
       txTr3,
       txTr4),
  from = 35671925,
  to = 35683156,
  transcriptAnnotation = "symbol"
)

```

# Dynamic Y-axis scaling

Gviz can overlay several DataTracks using the `OverlayTrack()` function, but the y-axis range is controlled by the first DataTrack. One must explicitly specify the y-axis range that is shared across all those overlayed tracks for proper comparison. The default behavior is to take the y-axis of the first DataTrack within the list of DataTracks being overlayed. This can be problematic if other DataTracks have different maximal signals, which results in a suboptimal y-axis range being set.

Here I write a function that can dynamically scale the Y-axis given one or more bigwigs that are plotted together. \## one sample

```{r dyanamic y-axis scaling function}
# Function to get max signal from bigWig in a given range
#' Title
#'
#' @param bigwig_file 
#' @param chr 
#' @param start 
#' @param end 
#'
#' @return
#' @export
#'
#' @examples
getMaxSignal <- function(bigwig_file, chr, start, end) {
  # Import specific genomic range from bigWig
  bw <- import(bigwig_file, 
              format = "bigWig",
              which = GRanges(seqnames = chr,
                             ranges = IRanges(start = start, 
                                            end = end)))
  
  # Return maximum signal value
  return(max(bw$score))
}

# Create and plot track with dynamic y-axis
createScaledTrack <- function(bigwig_file, chr, start, end, 
                             track_name = "Signal Track") {
  # Get maximum signal for scaling
  max_signal <- getMaxSignal(bigwig_file, chr, start, end)
  
  print(paste("The max signal for this bigwig is", max_signal))
  # Create DataTrack with adjusted y-limits
  dt <- DataTrack(range = bigwig_file,
                  chromosome = chr,
                  from = start,
                  to = end,
                  name = track_name,
                  type = "l",
                  ylim = c(0, max_signal * 1.1)) # Add 10% padding
  
  return(dt)
}
```

```{r testing dynamic y-axis scaling}
# Example usage
## comment out depending on PC or macbook:
 bigwig_file <- "data/ENCODE/ENCFF923TGK.bigWig"
#bigwig_file <- bwFile1

chr <- "chr9"
start <- 35671925
end <- 35683156


# Create track with automatic scaling
scaled_track <- createScaledTrack(bigwig_file, chr, start, end)

# Plot track
plotTracks(scaled_track, from = start, to = end)

# Combine with other tracks:
genome_axis <- GenomeAxisTrack()
plotTracks(list(genome_axis, scaled_track), from = start, to = end)

# compare unscaled to scaled
unscaled_track <- DataTrack(range = bigwig_file,
                  chromosome = chr,
                  from = start,
                  to = end,
                  name = "unscaled",
                  type = "l"
                  ) 

plotTracks(
  list(itrack,
       gtrack,
       unscaled_track,
       scaled_track,
       biomTrack1),
  from = 35671925,
  to = 35683156,
  col = c("red", "blue"),
  # set group colors
  transcriptAnnotation = "symbol"
)
```

Typically, more than one bigwig file is used to construct a figure. For example, if we are comparing control vs treatment or in the case below, normoxia vs hypoxia. Below we adjust the function to take in more than one input bigwig file: \## one or more samples

```{r one or more bw y-axis scaling}
# Function to get max signal from one or more bigWig files in a given range
getMaxSignal <- function(bigwig_files, chr, start, end) {
  # Handle single file input
  if (is.character(bigwig_files) && length(bigwig_files) == 1) {
    bigwig_files <- list(bigwig_files)
  }
  
  # Get maximum signal across all files
  max_signals <- sapply(bigwig_files, function(bw) {
    # Import specific genomic range from bigWig
    data <- import(bw,
                   format = "bigWig",
                   which = GRanges(
                     seqnames = chr,
                     ranges = IRanges(start = start,
                                      end = end)
                   ))
    
    if (length(data) == 0)
      return(0)
    return(max(data$score))
  })
  
  return(max(max_signals))
}

# Create and plot tracks with dynamic y-axis
createScaledTracks <- function(bigwig_files,
                               chr,
                               start,
                               end,
                               track_names = NULL,
                               colors = NULL,
                               padding_factor = 1.1,
                               overlay = "no") {
  # Handle input parameters
  if (is.character(bigwig_files) && length(bigwig_files) == 1) {
    bigwig_files <- list(bigwig_files)
  }
  
  # Generate default track names if not provided
  if (is.null(track_names)) {
    track_names <- paste("Signal", seq_along(bigwig_files))
  } else if (length(track_names) != length(bigwig_files)) {
    stop("Number of track names must match number of bigWig files")
  }
  
  # Generate default colors if not provided
  if (is.null(colors)) {
    colors <- rainbow(length(bigwig_files))
  } else if (length(colors) != length(bigwig_files)) {
    stop("Number of colors must match number of bigWig files")
  }
  
  # Get maximum signal across all files
  max_signal <- getMaxSignal(bigwig_files, chr, start, end)
  
  # Create DataTrack for each bigWig file
  tracks <- list()
  for (i in seq_along(bigwig_files)) {
    tracks[[i]] <- DataTrack(
      range = bigwig_files[[i]],
      chromosome = chr,
      from = start,
      to = end,
      name = track_names[i],
      type = "l",
      col = colors[i],
      # Explicitly assign track color by index
      baseline = 0,
      # Ensure histogram starts at 0
      col.baseline = "black",
      # set baseline color
      groups = factor(track_names[i],
                      levels = track_names),
     legend = TRUE,
      ylim = c(0, max_signal * padding_factor)
    )
  }
  
  # return scaled tracks:
  # option to overlay tracks
  if (tolower(overlay) == "yes") {
    tracks[[1]]@name <- "signal" # set shared y-axis title
    overlayTrack <- OverlayTrack(trackList = tracks)
    return(overlayTrack)
  } else{
    # default to stack tracks
    return(tracks)
  }
  
}

```

```{r example use of one or more samples}
# Example usage:

# Single file
single_bigwig <- "J:/genome-viewer/heatmap/RW555-1_treat_pileup.bw"
single_track <- createScaledTracks(single_bigwig, "chr9", 35671925, 35683156)

# Multiple files
bigwig_files <- list(
  #"J:/genome-viewer/heatmap/RW555-1_treat_pileup.bw",
  #"J:/genome-viewer/heatmap/RW555-2_treat_pileup.bw"
  "data/ENCODE/ENCFF923TGK.bigWig",
  "data/ENCODE/ENCFF923TGK.bigWig"
)

# Custom track names and colors
track_names <- c("normoxia", "hypoxia")
track_colors <- c("red", "blue")

# Create tracks with custom parameters
chr = "chr9"
start = 35671925
end = 35683156

scaledTracks <- createScaledTracks(
  bigwig_files = bigwig_files,
  chr = chr,
  start = start,
  end = end,
  track_names = track_names,
  colors = track_colors,
  padding_factor = 1.2  # 20% padding above max signal
)

# Plot tracks with genome axis
gtrack <- GenomeAxisTrack()
itrack <- IdeogramTrack(genome = "hg19", chromosome = chr)

plotTracks(c(list(itrack,
                  genome_axis), 
                  scaledTracks),
           legend = FALSE,
           from = start, to = end)

plotTracks(c(list(itrack,
                  gtrack),
             scaledTracks,
             biomTrack),
           from = start,
           to = end,
           #biomTrack1,
           col = c("red", "blue"),
           transcriptAnnotation = "symbol")

# test overlay
scaledTracks2 <- createScaledTracks(
  bigwig_files = bigwig_files,
  chr = chr,
  start = start,
  end = end,
  track_names = track_names,
  colors = track_colors,
  padding_factor = 1.2,  # 20% padding above max signal
  overlay = "yes"
)

plotTracks(
  c(list(itrack,
         gtrack),
    scaledTracks2,
    biomTrack1),
  from = start,
  to = end,
  background.title = "transparent",
  fontcolor.title = "black",
  # y-axis colors
  col.axis = "black",
  # y-axis colors
  col = c("red", "blue"),
  transcriptAnnotation = "symbol"
)


scaledTracks
```
##
```{r display parameter check, eval = FALSE}
displayPars(gtrack)

displayPars((scaledTracks[[1]]))
```

# BED annotation Track

Here I write a function that creates annotation tracks for bed files.

```{r bed annotation track}
# Create tracks with custom parameters
chr = "chr9"
start = 35671925
end = 35683156

# path to bed file(s)
bedFile <- "J:/genome-viewer/bed/2023-01-04_G361_N_72_vs_H_24_hypoxia_enrich.bed"
bedFile_many <- c("J:/genome-viewer/bed/2023-01-04_G361_N_72_vs_H_24_normoxia_enrich.bed",
                  "J:/genome-viewer/bed/2023-01-04_G361_N_72_vs_H_24_hypoxia_enrich.bed")

# import the bed as granges object
bed <- rtracklayer::import(bedFile)

# create annotationTrack
bedTrack <- AnnotationTrack(range = bed, genome = "hg19", 
                           name = "H24_peaks", chromosome = chr)
# plot
plotTracks(bedTrack, from = start, to = end)

# plot with previous tracks
plotTracks(
  c(list(itrack,
         gtrack),
    scaledTracks2,
    bedTrack,
    biomTrack1),
  from = start,
  to = end,
  background.title = "transparent",
  fontcolor.title = "black",
  # y-axis colors
  col.axis = "black",
  # y-axis colors
  col = c("red", "blue"),
  transcriptAnnotation = "symbol"
)
```

```{r}
create_bed_annotation_track <-
  function(bed_paths,
           track_names = NULL,
           color = "darkblue",
           chr) {
    # Input validation
    if (length(bed_paths) == 0) {
      stop("At least one BED file path must be provided")
    }
    
    # If no track names provided, use filenames
    if (is.null(track_names)) {
      track_names <- basename(bed_paths)
    }
    
    # Ensure track names match number of files
    if (length(track_names) != length(bed_paths)) {
      stop("Number of track names must match number of BED files")
    }
    
    # Ensure colors match number of track names
    if (length(color) != length(track_names)) {
      print("Number of colors should match number of BED files")
      # set to vector of darkblue
      color <- rep("darkblue",length(track_names))
    }
    
    # Create annotation tracks
    tracks <- list()
    for (i in seq_along(bed_paths)) {
      # Import BED file
      bed_data <- import(bed_paths[i])
      
      # Create Gviz annotation track
      tracks[[i]] <- AnnotationTrack(
        range = bed_data,
        chromosome = chr,
        name = track_names[i],
        color = color[i],
        fill = color[i]
       # stacking = "dense"
      )
    }
    
    return(tracks)
  }

# Example usage:
bedFile_many <-
  c(
    "J:/genome-viewer/bed/2023-01-04_G361_N_72_vs_H_24_normoxia_enrich.bed",
    "J:/genome-viewer/bed/2023-01-04_G361_N_72_vs_H_24_hypoxia_enrich.bed"
  )
track_names <- c("N_72", "H_24")
bed_colours <- c("darkred","darkblue")

bed_annotation_track <-
  create_bed_annotation_track(bedFile_many, track_names, bed_colours, chr = "chr9")

plotTracks(bed_annotation_track,
           from = start,
           to = end)


getOption("Gviz.scheme")
scheme <- getScheme()
scheme$AnnotationTrack$col <- "transparent"
scheme$GeneRegionTrack$col <- "transparent"
addScheme(scheme, "myScheme")
options(Gviz.scheme = "myScheme")

# plot with previous tracks
plotTracks(
  c(list(itrack,
         gtrack),
    scaledTracks2,
    bed_annotation_track,
   biomTrack1
    ),
  from = start,
  to = end,
  background.title = "transparent",
  fontcolor.title = "black",
  col.axis = "black",
  col = c("green", "red"),  # y-axis colors
  transcriptAnnotation = "symbol"
)

pdf("test.pdf")
plotTracks(
  c(list(itrack,
         gtrack),
    scaledTracks2,
    bed_annotation_track,
   biomTrack1
    ),
  from = start,
  to = end,
  background.title = "transparent",
  fontcolor.title = "black",
  col.axis = "black",
  #col = c("green", "red"),  # y-axis colors
  transcriptAnnotation = "symbol"
)
dev.off()
```

```         
txTr <- GeneRegionTrack(txdb,
                        chromosome = chr,
                        start = 35000000,
                        end = 40000000)
```
